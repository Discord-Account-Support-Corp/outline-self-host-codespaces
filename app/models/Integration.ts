import { observable } from "mobx";
import {
  IntegrationService,
  MentionType,
  type IntegrationSettings,
  type IntegrationType,
} from "@shared/types";
import User from "~/models/User";
import Model from "~/models/base/Model";
import Field from "~/models/decorators/Field";
import Relation from "~/models/decorators/Relation";

class Integration<T = unknown> extends Model {
  static modelName = "Integration";

  type: IntegrationType;

  service: IntegrationService;

  collectionId: string;

  userId: string;

  @Relation(() => User, { onDelete: "cascade" })
  user: User;

  @Field
  @observable
  events: string[];

  @observable
  settings: IntegrationSettings<T>;

  isMentionable(url: URL) {
    const { hostname, pathname } = url;
    const pathParts = pathname.split("/");

    switch (this.service) {
      case IntegrationService.GitHub: {
        const settings = this
          .settings as IntegrationSettings<IntegrationType.Embed>;

        return (
          hostname === "github.com" &&
          settings.github?.installation.account.name === pathParts[1] // ensure installed org/account name matches with the provided url.
        );
      }

      default:
        return false;
    }
  }

  getMentionType(url: URL): MentionType | undefined {
    const { pathname } = url;
    const pathParts = pathname.split("/");

    switch (this.service) {
      case IntegrationService.GitHub: {
        const type = pathParts[3];
        return type === "pull"
          ? MentionType.PullRequest
          : type === "issues"
          ? MentionType.Issue
          : undefined;
      }

      default:
        return;
    }
  }
}

export default Integration;
