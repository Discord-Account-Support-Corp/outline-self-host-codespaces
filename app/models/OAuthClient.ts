import { observable } from "mobx";
import User from "./User";
import ParanoidModel from "./base/ParanoidModel";
import Field from "./decorators/Field";
import Relation from "./decorators/Relation";

class OAuthClient extends ParanoidModel {
  static modelName = "OAuthClient";

  /** The human-readable name of this app */
  @Field
  @observable
  name: string;

  /** A short description of this app */
  @Field
  @observable
  description: string | null;

  /** The name of the developer of this app */
  @Field
  @observable
  developerName: string | null;

  /** The URL of the developer of this app */
  @Field
  @observable
  developerUrl: string | null;

  /** The URL of the avatar of the developer of this app */
  @Field
  @observable
  avatarUrl: string | null;

  /** The developer facing ID of this app */
  @Field
  clientId: string;

  /** The developer facing secret of this app */
  @Field
  clientSecret: string;

  /** Whether this app is published (available to other workspaces) */
  @Field
  @observable
  published: boolean;

  /** A list of valid redirect URIs for this app */
  @Field
  @observable
  redirectUris: string[];

  @Relation(() => User)
  createdBy: User;

  createdById: string;
}

export default OAuthClient;
