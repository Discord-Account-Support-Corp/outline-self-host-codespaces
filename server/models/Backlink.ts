import { InferAttributes, InferCreationAttributes } from "sequelize";
import {
  DataType,
  BelongsTo,
  ForeignKey,
  Column,
  Table,
} from "sequelize-typescript";
import Document from "./Document";
import Relationship, { RelationshipType } from "./Relationship";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";

@Table({ tableName: "backlinks", modelName: "backlink" })
@Fix
class Backlink extends IdModel<
  InferAttributes<Backlink>,
  Partial<InferCreationAttributes<Backlink>>
> {
  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  userId: string;

  @BelongsTo(() => Document, "documentId")
  document: Document;

  @ForeignKey(() => Document)
  @Column(DataType.UUID)
  documentId: string;

  @BelongsTo(() => Document, "reverseDocumentId")
  reverseDocument: Document;

  @ForeignKey(() => Document)
  @Column(DataType.UUID)
  reverseDocumentId: string;

  /**
   * Find all backlinks for a document that the user has access to
   *
   * @param documentId The document ID to find backlinks for
   * @param user The user to check access for
   */
  public static async findSourceDocumentIdsForUser(
    documentId: string,
    user: User
  ) {
    // Delegate to Relationship model for actual implementation
    return Relationship.findSourceDocumentIdsForUser(documentId, user);
  }

  /**
   * Create a new backlink relationship
   */
  public static async findOrCreate(options: any) {
    // Delegate to Relationship model with backlink type
    return Relationship.findOrCreate({
      ...options,
      defaults: {
        ...options.defaults,
        type: RelationshipType.Backlink,
      },
    });
  }

  /**
   * Find all backlinks matching criteria
   */
  public static async findAll(options: any) {
    // Delegate to Relationship model with backlink type filter
    return Relationship.findAll({
      ...options,
      where: {
        ...options.where,
        type: RelationshipType.Backlink,
      },
    });
  }

  /**
   * Destroy backlinks matching criteria
   */
  public static async destroy(options: any) {
    // Delegate to Relationship model with backlink type filter
    return Relationship.destroy({
      ...options,
      where: {
        ...options.where,
        type: RelationshipType.Backlink,
      },
    });
  }
}

export default Backlink;
