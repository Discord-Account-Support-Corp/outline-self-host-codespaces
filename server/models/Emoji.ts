import { InferAttributes, InferCreationAttributes } from "sequelize";
import {
  BelongsTo,
  Column,
  DataType,
  ForeignKey,
  Index,
  Table,
} from "sequelize-typescript";
import { EmojiValidation } from "@shared/validations";
import Team from "./Team";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";
import Length from "./validators/Length";

@Table({ tableName: "emojis", modelName: "emoji" })
@Fix
class Emoji extends IdModel<
  InferAttributes<Emoji>,
  Partial<InferCreationAttributes<Emoji>>
> {
  @Length({
    max: EmojiValidation.maxNameLength,
    msg: `name must be ${EmojiValidation.maxNameLength} characters or less`,
  })
  @Column(DataType.STRING)
  name: string;

  @Length({
    max: EmojiValidation.maxUrlLength,
    msg: `url must be ${EmojiValidation.maxUrlLength} characters or less`,
  })
  @Column(DataType.STRING)
  url: string;

  // associations

  @BelongsTo(() => Team)
  team: Team;

  @ForeignKey(() => Team)
  @Index
  @Column(DataType.UUID)
  teamId: string;

  @BelongsTo(() => User)
  createdBy: User;

  @ForeignKey(() => User)
  @Index
  @Column(DataType.UUID)
  createdById: string;
}

export default Emoji;
