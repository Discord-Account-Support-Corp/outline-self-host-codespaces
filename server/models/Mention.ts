import { InferAttributes, InferCreationAttributes } from "sequelize";
import {
  DataType,
  BelongsTo,
  ForeignKey,
  Column,
  Table,
} from "sequelize-typescript";
import { MentionType } from "@shared/types";
import { can } from "@server/policies";
import Document from "./Document";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";

@Table({ tableName: "mentions", modelName: "mention" })
@Fix
class Mention extends IdModel<
  InferAttributes<Mention>,
  Partial<InferCreationAttributes<Mention>>
> {
  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  userId: string;

  @BelongsTo(() => Document, "documentId")
  document: Document;

  @ForeignKey(() => Document)
  @Column(DataType.UUID)
  documentId: string;

  @BelongsTo(() => User, "mentionedUserId")
  mentionedUser: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  mentionedUserId: string;

  @Column(DataType.STRING)
  mentionType: MentionType;

  @Column(DataType.STRING)
  mentionId: string;

  /**
   * Find all mentions for a user in documents they have access to
   *
   * @param userId The user ID to find mentions for
   * @param user The user to check document access for
   */
  public static async findMentionsForUser(userId: string, user: User) {
    const mentions = await this.findAll({
      where: {
        mentionedUserId: userId,
      },
      include: [
        {
          model: Document,
          as: "document",
        },
      ],
    });

    // Filter mentions to only include documents the user has access to
    const accessibleMentions = [];
    for (const mention of mentions) {
      if (mention.document) {
        const hasAccess = can(user, "read", mention.document);
        if (hasAccess) {
          accessibleMentions.push(mention);
        }
      }
    }

    return accessibleMentions;
  }

  /**
   * Find all mentions in a specific document
   *
   * @param documentId The document ID to find mentions for
   */
  public static async findMentionsInDocument(documentId: string) {
    return this.findAll({
      where: {
        documentId,
      },
      include: [
        {
          model: User,
          as: "mentionedUser",
        },
      ],
    });
  }
}

export default Mention;
