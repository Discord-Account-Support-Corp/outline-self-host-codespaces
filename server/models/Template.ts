import { InferAttributes, InferCreationAttributes, Op } from "sequelize";
import {
  Column,
  DataType,
  BelongsTo,
  ForeignKey,
  IsDate,
  Table,
  Length as SimpleLength,
  DefaultScope,
  Default,
  HasMany,
  Unique,
} from "sequelize-typescript";
import slugify from "slugify";
import { ProsemirrorData } from "@shared/types";
import { DocumentValidation } from "@shared/validations";
import Collection from "./Collection";
import Revision from "./Revision";
import Team from "./Team";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";
import IsHexColor from "./validators/IsHexColor";
import Length from "./validators/Length";

@DefaultScope(() => ({
  include: [
    {
      model: User,
      as: "createdBy",
      paranoid: false,
    },
    {
      model: User,
      as: "updatedBy",
      paranoid: false,
    },
  ],
  where: {
    publishedAt: {
      [Op.ne]: null,
    },
    template: true,
  },
  attributes: {
    exclude: ["state"],
  },
}))
@Table({ tableName: "documents", modelName: "template" })
@Fix
class Template extends IdModel<
  InferAttributes<Template>,
  Partial<InferCreationAttributes<Template>>
> {
  @SimpleLength({
    min: 10,
    max: 10,
    msg: `urlId must be 10 characters`,
  })
  @Unique
  @Column
  urlId: string;

  @Length({
    max: DocumentValidation.maxTitleLength,
    msg: `Template title must be ${DocumentValidation.maxTitleLength} characters or less`,
  })
  @Column
  title: string;

  @Default(false)
  @Column
  fullWidth: boolean;

  /** The version of the editor last used to edit this document. */
  @SimpleLength({
    max: 255,
    msg: `editorVersion must be 255 characters or less`,
  })
  @Column
  editorVersion: string;

  /** An icon to use as the template icon. */
  @Length({
    max: 50,
    msg: `icon must be 50 characters or less`,
  })
  @Column
  icon: string | null;

  /** The color of the icon. */
  @IsHexColor
  @Column
  color: string | null;

  /**
   * The content of the template as JSON, this is a snapshot at the last time the state was saved.
   */
  @Column(DataType.JSONB)
  content: ProsemirrorData | null;

  /** Whether the document is published, and if so when. */
  @IsDate
  @Column
  publishedAt: Date | null;

  // associations

  @BelongsTo(() => User, "lastModifiedById")
  updatedBy: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  lastModifiedById: string;

  @BelongsTo(() => User, "createdById")
  createdBy: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  createdById: string;

  @BelongsTo(() => Team, "teamId")
  team: Team;

  @ForeignKey(() => Team)
  @Column(DataType.UUID)
  teamId: string;

  @BelongsTo(() => Collection, "collectionId")
  collection: Collection | null;

  @ForeignKey(() => Collection)
  @Column(DataType.UUID)
  collectionId?: string | null;

  @HasMany(() => Revision, "documentId")
  revisions: Revision[];

  @Default(true)
  @Column
  template: boolean;

  // getters

  /** The frontend path to this template. */
  get path() {
    if (!this.title) {
      return `/settings/templates/untitled-${this.urlId}`;
    }
    const slugifiedTitle = slugify(this.title);
    return `/settings/templates/${slugifiedTitle}-${this.urlId}`;
  }

  /**
   * Returns whether this is a workspace template.
   *
   * @returns boolean
   */
  get isWorkspaceTemplate() {
    return !this.collectionId;
  }

  /**
   * Convenience method that returns whether this template is a draft.
   *
   * @returns boolean
   */
  get isDraft(): boolean {
    return !this.publishedAt;
  }

  /**
   * Convenience method that returns whether this template is deleted.
   *
   * @returns boolean
   */
  get isDeleted(): boolean {
    return !this.deletedAt;
  }
}

export default Template;
