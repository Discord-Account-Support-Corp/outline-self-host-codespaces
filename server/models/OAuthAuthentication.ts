import crypto from "crypto";
import { addDays, addHours, subMinutes } from "date-fns";
import rs from "randomstring";
import { InferAttributes, InferCreationAttributes } from "sequelize";
import {
  Column,
  DataType,
  BelongsTo,
  ForeignKey,
  Table,
  BeforeCreate,
  IsDate,
  Unique,
} from "sequelize-typescript";
import OAuthClient from "./OAuthClient";
import User from "./User";
import ParanoidModel from "./base/ParanoidModel";
import { SkipChangeset } from "./decorators/Changeset";
import Fix from "./decorators/Fix";

@Table({
  tableName: "oauth_authentications",
  modelName: "oauth_authentication",
})
@Fix
class OAuthAuthentication extends ParanoidModel<
  InferAttributes<OAuthAuthentication>,
  Partial<InferCreationAttributes<OAuthAuthentication>>
> {
  public static accessTokenPrefix = "ol_atx_";
  public static refreshTokenPrefix = "ol_rtx_";

  @Unique
  @Column
  @SkipChangeset
  accessTokenHash: string;

  /** The cached plain text access token. Only available during creation. */
  @Column(DataType.VIRTUAL)
  accessToken: string | null;

  @IsDate
  @Column
  accessTokenExpiresAt: Date;

  @Unique
  @Column
  @SkipChangeset
  refreshTokenHash: string;

  /** The cached plain text refresh token. Only available during creation. */
  @Column(DataType.VIRTUAL)
  refreshToken: string | null;

  @IsDate
  @Column
  refreshTokenExpiresAt: Date;

  @Column
  scope: string;

  @IsDate
  @Column
  @SkipChangeset
  lastActiveAt: Date;

  // associations

  @BelongsTo(() => OAuthClient, "oauthClientId")
  oauthClient: OAuthClient;

  @ForeignKey(() => OAuthClient)
  @Column(DataType.UUID)
  oauthClientId: string;

  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  userId: string;

  // methods

  updateActiveAt = async () => {
    const fiveMinutesAgo = subMinutes(new Date(), 5);

    // ensure this is updated only every few minutes otherwise
    // we'll be constantly writing to the DB as API requests happen
    if (!this.lastActiveAt || this.lastActiveAt < fiveMinutesAgo) {
      this.lastActiveAt = new Date();
    }

    return this.save({ silent: true });
  };

  // hooks

  @BeforeCreate
  static async generateTokens(model: OAuthAuthentication) {
    const accessToken = `${this.accessTokenPrefix}${rs.generate(32)}`;
    model.accessToken = accessToken;
    model.accessTokenHash = this.hash(accessToken);
    model.accessTokenExpiresAt = addHours(new Date(), 1);

    const refreshToken = `${this.refreshTokenPrefix}${rs.generate(32)}`;
    model.refreshToken = refreshToken;
    model.refreshTokenHash = this.hash(refreshToken);
    model.refreshTokenExpiresAt = addDays(new Date(), 30);
  }

  @BeforeCreate
  static async setLastActiveAt(model: OAuthAuthentication) {
    model.lastActiveAt = new Date();
  }

  // static methods

  /**
   * Generates a hashed token for the given input.
   *
   * @param key The input string to hash
   * @returns The hashed input
   */
  public static hash(key: string) {
    return crypto.createHash("sha256").update(key).digest("hex");
  }

  /**
   * Finds an OAuthAuthentication by the given input string.
   *
   * @param input The input string to search for
   * @returns The OAuthAuthentication if found
   */
  public static findByAccessToken(input: string) {
    return this.findOne({
      where: {
        accessTokenHash: this.hash(input),
      },
    });
  }
}

export default OAuthAuthentication;
