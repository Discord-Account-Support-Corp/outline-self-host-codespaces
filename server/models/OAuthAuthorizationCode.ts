import { InferAttributes, InferCreationAttributes } from "sequelize";
import {
  Column,
  DataType,
  BelongsTo,
  ForeignKey,
  Table,
  Length,
} from "sequelize-typescript";
import { OAuthClientValidation } from "@shared/validations";
import OAuthClient from "./OAuthClient";
import User from "./User";
import IdModel from "./base/IdModel";
import Fix from "./decorators/Fix";

@Table({
  tableName: "oauth_authorization_codes",
  modelName: "oauth_authorization_code",
  timestamps: false,
})
@Fix
class OAuthAuthorizationCode extends IdModel<
  InferAttributes<OAuthAuthorizationCode>,
  Partial<InferCreationAttributes<OAuthAuthorizationCode>>
> {
  @Column
  authorizationCodeHash: string;

  @Column
  scope: string;

  @Length({ max: OAuthClientValidation.maxRedirectUriLength })
  @Column
  redirectUri: string;

  @Column(DataType.DATE)
  expiresAt: Date;

  @Column(DataType.DATE)
  createdAt: Date;

  // associations

  @BelongsTo(() => OAuthClient, "oauthClientId")
  oauthClient: OAuthClient;

  @ForeignKey(() => OAuthClient)
  @Column(DataType.UUID)
  oauthClientId: string;

  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column(DataType.UUID)
  userId: string;
}

export default OAuthAuthorizationCode;
