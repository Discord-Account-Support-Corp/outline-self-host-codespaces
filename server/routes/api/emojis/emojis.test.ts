import { v4 } from "uuid";
import { Emoji } from "@server/models";
import {
  buildAdmin,
  buildEmoji,
  buildTeam,
  buildUser,
} from "@server/test/factories";
import { getTestServer } from "@server/test/support";

const server = getTestServer();

describe("#emojis.info", () => {
  it("should return emoji info by name", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const emoji = await buildEmoji({
      teamId: team.id,
      createdById: user.id,
      name: "testemoji",
    });

    const res = await server.post("/api/emojis.info", {
      body: {
        token: user.getJwtToken(),
        name: emoji.name,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(emoji.id);
    expect(body.data.name).toEqual(emoji.name);
    expect(body.data.url).toEqual(emoji.url);
  });

  it("should require authentication", async () => {
    const res = await server.post("/api/emojis.info");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should return emoji info", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const emoji = await buildEmoji({
      teamId: team.id,
      createdById: user.id,
    });

    const res = await server.post("/api/emojis.info", {
      body: {
        token: user.getJwtToken(),
        id: emoji.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(emoji.id);
    expect(body.data.name).toEqual(emoji.name);
    expect(body.data.url).toEqual(emoji.url);
  });

  it("should not return emoji from another team", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const otherTeam = await buildTeam();
    const otherUser = await buildUser({ teamId: otherTeam.id });
    const emoji = await buildEmoji({
      teamId: otherTeam.id,
      createdById: otherUser.id,
    });

    const res = await server.post("/api/emojis.info", {
      body: {
        token: user.getJwtToken(),
        id: emoji.id,
      },
    });

    expect(res.status).toEqual(403);
  });
});

describe("#emojis.list", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/emojis.list");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should return team emojis", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    await buildEmoji({
      teamId: team.id,
      createdById: user.id,
      name: "awesome",
    });
    await buildEmoji({
      teamId: team.id,
      createdById: user.id,
      name: "cool",
    });

    // Create emoji in another team that should not be returned
    const otherTeam = await buildTeam();
    const otherUser = await buildUser({ teamId: otherTeam.id });
    await buildEmoji({
      teamId: otherTeam.id,
      createdById: otherUser.id,
      name: "hidden",
    });

    const res = await server.post("/api/emojis.list", {
      body: {
        token: user.getJwtToken(),
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data).toHaveLength(2);
    expect(body.data.map((e: Emoji) => e.name).sort()).toEqual([
      "awesome",
      "cool",
    ]);
    expect(body.data[0].createdBy).toBeDefined();
    expect(body.policies).toBeDefined();
  });

  it("should support pagination", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    // Create multiple emojis
    for (let i = 0; i < 15; i++) {
      await buildEmoji({
        teamId: team.id,
        createdById: user.id,
        name: `emoji${i}`,
      });
    }

    const res = await server.post("/api/emojis.list", {
      body: {
        token: user.getJwtToken(),
        limit: 10,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data).toHaveLength(10);
    expect(body.pagination.total).toEqual(15);
  });
});

describe("#emojis.create", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/emojis.create");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should create emoji", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user.getJwtToken(),
        name: "awesome",
        url: "https://example.com/awesome.png",
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.name).toEqual("awesome");
    expect(body.data.url).toEqual("https://example.com/awesome.png");
    expect(body.data.createdBy.id).toEqual(user.id);
    expect(body.policies).toBeDefined();

    // Verify emoji was created in database
    const emoji = await Emoji.findByPk(body.data.id);
    expect(emoji).toBeTruthy();
    expect(emoji!.teamId).toEqual(team.id);
    expect(emoji!.createdById).toEqual(user.id);
  });

  it("should not allow duplicate emoji names in same team", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    await buildEmoji({
      teamId: team.id,
      createdById: user.id,
      name: "awesome",
    });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user.getJwtToken(),
        name: "awesome",
        url: "https://example.com/awesome2.png",
      },
    });

    expect(res.status).toEqual(400);
    const body = await res.json();
    expect(body.message).toContain("already exists");
  });

  it("should allow same emoji name in different teams", async () => {
    const team1 = await buildTeam();
    const user1 = await buildUser({ teamId: team1.id });
    const team2 = await buildTeam();
    const user2 = await buildUser({ teamId: team2.id });

    await buildEmoji({
      teamId: team1.id,
      createdById: user1.id,
      name: "awesome",
    });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user2.getJwtToken(),
        name: "awesome",
        url: "https://example.com/awesome2.png",
      },
    });

    expect(res.status).toEqual(200);
    const body = await res.json();
    expect(body).toMatchObject({
      data: expect.objectContaining({
        name: "awesome",
      }),
    });
  });

  it("should validate emoji name length", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user.getJwtToken(),
        name: "a".repeat(51), // Exceeds max length of 50
        url: "https://example.com/awesome.png",
      },
    });

    expect(res.status).toEqual(400);
  });

  it("should validate emoji URL length", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user.getJwtToken(),
        name: "awesome",
        url: "https://example.com/" + "a".repeat(500), // Exceeds max length of 500
      },
    });

    expect(res.status).toEqual(400);
  });

  it("should validate URL format", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    const res = await server.post("/api/emojis.create", {
      body: {
        token: user.getJwtToken(),
        name: "awesome",
        url: "not-a-valid-url",
      },
    });

    expect(res.status).toEqual(400);
  });
});

describe("#emojis.delete", () => {
  it("should require authentication", async () => {
    const res = await server.post("/api/emojis.delete");
    const body = await res.json();
    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it("should allow emoji creator to delete", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });
    const emoji = await buildEmoji({
      teamId: team.id,
      createdById: user.id,
    });

    const res = await server.post("/api/emojis.delete", {
      body: {
        token: user.getJwtToken(),
        id: emoji.id,
      },
    });

    expect(res.status).toEqual(200);
    const deleteBody = await res.json();
    expect(deleteBody).toMatchObject({
      success: true,
    });

    // Verify emoji was deleted
    const deletedEmoji = await Emoji.findByPk(emoji.id);
    expect(deletedEmoji).toBeNull();
  });

  it("should allow team admin to delete any emoji", async () => {
    const team = await buildTeam();
    const admin = await buildAdmin({ teamId: team.id });
    const user = await buildUser({ teamId: team.id });
    const emoji = await buildEmoji({
      teamId: team.id,
      createdById: user.id,
    });

    const res = await server.post("/api/emojis.delete", {
      body: {
        token: admin.getJwtToken(),
        id: emoji.id,
      },
    });

    expect(res.status).toEqual(200);
    const adminDeleteBody = await res.json();
    expect(adminDeleteBody).toMatchObject({
      success: true,
    });

    // Verify emoji was deleted
    const deletedEmoji = await Emoji.findByPk(emoji.id);
    expect(deletedEmoji).toBeNull();
  });

  it("should not allow non-creator non-admin to delete", async () => {
    const team = await buildTeam();
    const creator = await buildUser({ teamId: team.id });
    const otherUser = await buildUser({ teamId: team.id });
    const emoji = await buildEmoji({
      teamId: team.id,
      createdById: creator.id,
    });

    const res = await server.post("/api/emojis.delete", {
      body: {
        token: otherUser.getJwtToken(),
        id: emoji.id,
      },
    });

    expect(res.status).toEqual(403);

    // Verify emoji was not deleted
    const stillExists = await Emoji.findByPk(emoji.id);
    expect(stillExists).toBeTruthy();
  });

  it("should not allow deleting emoji from another team", async () => {
    const team1 = await buildTeam();
    const user1 = await buildUser({ teamId: team1.id });
    const team2 = await buildTeam();
    const user2 = await buildUser({ teamId: team2.id });
    const emoji = await buildEmoji({
      teamId: team2.id,
      createdById: user2.id,
    });

    const res = await server.post("/api/emojis.delete", {
      body: {
        token: user1.getJwtToken(),
        id: emoji.id,
      },
    });

    expect(res.status).toEqual(403);

    // Verify emoji was not deleted
    const stillExists = await Emoji.findByPk(emoji.id);
    expect(stillExists).toBeTruthy();
  });

  it("should return 404 for non-existent emoji", async () => {
    const team = await buildTeam();
    const user = await buildUser({ teamId: team.id });

    const res = await server.post("/api/emojis.delete", {
      body: {
        token: user.getJwtToken(),
        id: v4(), // Non-existent UUID
      },
    });

    expect(res.status).toEqual(404);
  });
});
