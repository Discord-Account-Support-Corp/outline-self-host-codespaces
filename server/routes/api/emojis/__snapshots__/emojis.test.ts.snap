// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emojis.create should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#emojis.delete should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#emojis.info should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#emojis.list should require authentication 1`] = `
{
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;
